### ISSUE TITLE
Make the ActionDispatch::Request class methods available in views

### ISSUE THREAD

**Issue by evenreven** [phlex-ruby/phlex#248]
Not sure if this is feasible or if it would even be a good thing... 

In my quest to recreate an app's view layer with Phlex, I'm trying to use the `request.original_url` method to put in the `head`, i.e. `<link rel="canonical" href="https://www.mywebsite.com/page">`. I'm trying to do this with `link(rel: "canonical", href: request.original_url)`, but trying to add `request` with `raw helpers` doesn't work, and in fact it breaks the entire head tag. I've unsuccessfully tried to include the relevant ActionDispatch module, but failed miserably since request is a class and I don't know how to work around this. 

I could get the original_url in the controller instead and pass it as an argument to the view, but it seems cumbersome to do this in every controller, since that's what I would have to do to pass it to the Layout (unless I'm mistaken). 

Would it make sense to make `request` available to views (as it is in ActionView), or should I find a different way to do this? (It could also very well be that this is already possible, but that I'm just doing it wrong! If so, just close and disregard.)

**Comment by joeldrapper** [comment]
You should be able to access the `request` through the `helpers` proxy. Exposing `request` directly in Phlex views is an interesting idea.

**Comment by evenreven** [comment]
I think maybe my syntax when calling the helper is off. This is what I'm doing, and it does not work:

```ruby
module Views
  class Layout < ApplicationView
    def initialize(title:)
      @title = title
    end

    def template(&)
      doctype

      html do
        head do
          link rel: "canonical", href: original_url
        end

        body do
          main(&)
        end
      end
    end

    private

    def original_url
      raw helpers.request.original_url
    end
  end
end
```

The original url gets output, so it's working in a technical sense, but instead of staying inside the block, it "overflows" (not the correct term, I'm sure) so that the url is suddenly outside the link element and the url is visible on the page. The browser inspector says it's in the body tag on its own, outside of the link rel tag. Any further tags in head after link rel gets pushed into the body too. This is the HTML, however, from view source:

```html
<!DOCTYPE html>
<html>
  <head>http://localhost:3000/study_programmes/1
    <link rel="canonical" />
    <title>Page title</title>
  </head>

  <body>
    <main>
      <h1>Page title</h1>
      <!-- Snip -->
    </main>
  </body>
</html>
```

Apart from the fact the inspector and view source disagree on the hierarchy, it's still in the wrong place. Any ideas? 

(Setting request.original_url on an instance variable in application_controller and passing that along from the study_programmes controller works well, but is cumbersome since I only really need it in the layout.)

**Comment by joeldrapper** [comment]
`raw` immediately outputs the string passed to. If you remove `raw`, this should work.

**Comment by evenreven** [comment]
🤦‍♂️

I was so obsessed with getting the request call right that it didn't occur to me that the problem was `raw`. It's working (obviously!), sorry for wasting your time! Closing.

**Comment by johnmcdowall** [comment]
@joeldrapper Sorry to necro this thread, but here's an important use case as to why making `request` natively available within scope: 

I have some helpers in `ApplicationHelper` that rely on being able to access the `request` to be able to make some decisions based on path. 

These helpers are used throughout the App which is migrating to Phlex so has vary many partials that just expect these helpers to work. 

If I try to use those helpers within a Phlex component, it breaks because `request` isn't accessible. The inverse is also a problem: the `helpers` attribute isn't accessible inside an ordinary helper. 

So what I've done is the following (placed at the start of the helper method in question), which is a hack and not great: 

```
request = if defined?(helpers)
  helpers.request
else
  @_request
end
```

I wouldn't need to do this if `request` was natively within the scope of a Phlex component.

**Comment by joeldrapper** [comment]
@johnmcdowall I would accept a PR defining a request as a value helper adapter like this one. https://github.com/phlex-ruby/phlex-rails/blob/71db8cc0b586eb0600a4dfb146d71a6ecacaeacb/lib/phlex/rails/helpers.rb#L44-L50

**Comment by kaspth** [comment]
@johnmcdowall you could potentially add a `delegate :request, to: :helpers` in your `ApplicationView` (or whatever your common ancestor for Phlex views is called) class too as a workaround to help the conversion.

**Comment by joeldrapper** [comment]
> you could potentially add a delegate :request, to: :helpers in your ApplicationView (or whatever your common ancestor for Phlex views is called) class too as a workaround to help the conversion.

Yes, for value helpers, delegating does exactly the same thing. It would be nice to have an adapter with the others though. The advantage of the adapters is you don't need to spend any thought on what kind of helper it is — value, output, builder-yielding-output, etc, you just add the adapter and use it naturally.
